### ISSUE TITLE
microk8s behind corporate proxy needs extra configuration besides /etc/environment

### ISSUE THREAD

**Issue by petermicuch** [canonical/microk8s.io#576]
## Summary

In the [article](https://microk8s.io/docs/install-proxy), specifically the part **Configuring containerd only**, there is incorrect statement that "_Containerd uses the proxy configuration set in /etc/environment, therefore no extra configuration is typically required._" This is not true. When running behind corporate proxy, you have to set `HTTP/HTTPS_PROXY` and `NO_PROOXY` variables respectively in the ${SNAP_DATA}/args/containerd-env and other files (kube-apiserver, kube-proxy). Otherwise containers do not get pulled, etc. I am not the first one to hit the problem as you can see [here](https://github.com/canonical/microk8s/issues/344#issuecomment-469150203) and if you search for "microk8s proxy" you will find bunch of threads where people got confused.

Also the content of the `NO_PROXY` variable is not very clear. Usually those who install microk8s behind proxy do know, what are their internal IP ranges/hosts. But is it really necessary to specify whole 10.0.0.0/8? Could we be maybe a bit more specific, what exact ranges are used by microk8s?

## Process

Read the referenced article and try to install microk8s on the machine behind the proxy server.

## Current and expected result

Update\Reformulate the text, that setting of the proxy information is mandatory also for files in `${SNAP_DATA}/args/` directory and that using `/etc/environment` is not sufficient. Consider to elaborate on variable ${SNAP_DATA} as it is not set by default on the installed systems (I could not find it). 

Thanks a lot for considering this :-)

## Screenshot

n.a.

## Browser details

I was testing this on Red Hat Enterprise Linux release 8.6 (Ootpa) and MicroK8s v1.26.0 revision 4390

**Comment by petesfrench** [comment]
Hi, thank you for reporting this, we will work on it ASAP

**Comment by petesfrench** [comment]
@evilnick Is this something you could look at?

**Comment by petermicuch** [comment]
@petesfrench thank you for taking care.

I would add a note to the NO_PROXY just for more clarity (I hope). Right now there is [example](https://microk8s.io/docs/install-proxy) that suggests to use `NO_PROXY=10.0.0.0/8,192.168.0.0/16,127.0.0.1,172.16.0.0/16`. I am extending that one with values `localhost,127.0.0.1,.svc,.svc.cluster.local` that I always keep in my no proxy in order to not break in-cluster communication. What is the bare minimum that I must have inside `NO_PROXY` for microk8s to be able to configure `HTTP_PROXY` and `HTTPS_PROXY` as an environment for a pod, that I would like to allow to communicate with outside world?

I guess setting of only `HTTPS_PROXY` without any `NO_PROXY` being set is fine, as long as communication inside cluster is pure `HTTP`. But let's say I want to access some insecure `HTTP` only site (yeah... I know :-)), how do I then correctly set `NO_PROXY`? Not sure if answer to this question shall bundled with this issue itself or rather be part of troubleshooting.

**Comment by nanoric** [comment]
For microk8s 1.26.7, ```HTTP_PROXY``` and ```HTTPS_PROXY``` in ```containerd-env``` has no effect, only ```NO_PROXY``` is useful.
So for me:
 * In order to perform a successful image pulling, set ```HTTP_PROXY``` and ```HTTPS_PROXY``` in ```/etc/environment```.
 *  then set ```NO_PROXY``` in ```/var/snap/microk8s/current/args/containerd-env```.
 * In order to communicate to systemd's dns server, set ```127.0.0.1```  in ```NO_PROXY```.
 * For in-cluster communication, set ```*.svc,*.cluster.local``` ([refering_this](https://github.com/cert-manager/cert-manager/issues/2640#issuecomment-601872165)) and ```10.0.0.0/8,172.16.0.0/16,192.168.0.0/16``` in ```NO_PROXY```.

Also, some people, like me, don't like to use ```/etc/environment``` because it is a global setting: not only microk8s will be effected, but other program that don't need a proxy will be effected. Maybe ```HTTP_PROXY``` and ```HTTPS_PROXY``` in ```containerd-env```  is a bug in document/microk8s? Can it be fixed?

**Comment by petermicuch** [comment]
@evilnick - wov, I have note realized this issue will soon have its anniversary. Was there anything concluded by the team?

**Comment by petermicuch** [comment]
Hi @ktsakalozos 

I am aware of the global `/etc/environment`, what I was trying to say is that even when that was configured it was not sufficient. Unless I configured proxy settings in ${SNAP_DATA}/args/containerd-env containers were not pulled. You are only able to reproduce this in environment which blocks any requests that does not go over the proxy (i.e. properly configured firewall for outgoing traffic). If you think update of the https://microk8s.io/docs/install-proxy is not necessary I am fine with that. Closing this issue now.
